File: solns/scala/67.scala
import scala.io.Source

val data = Source.fromFile("../data/67.txt")
    .getLines
    .toList
    .map(_.trim.split(" ").map(_.toLong))
    .toArray

for (i <- data.length - 2 to 0 by -1;
     j <- 0 until data(i).length) {
  val left = data(i + 1)(j)
  val right = data(i + 1)(j + 1)
  data(i)(j) += left max right
}

val result = data(0)(0)

println(result)
assert(result == 7273)

File: solns/ruby/67.rb
data = []
file = File.new('/data/67.txt')
while (line = file.gets)
  data << line
end
file.close

data = data.map(&:strip)
           .reject { |l| l == '' }
           .map { |l| l.split.map(&:to_i) }

(data.length - 2).step(0, -1).each do |i|
  (0...data[i].length).each do |j|
    left = data[i + 1][j]
    right = data[i + 1][j + 1]
    data[i][j] += [left, right].max
  end
end

result = data[0][0]

puts result
raise Error unless result == 7273
